### PR TITLE
Refine DB manager for state migration

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes
This PR refine DB manager for state migration.

- add `stateMigrationDB`
- add methods for old/new stateDB
- chang order of DBs (MiscDB is the 1st DB)
- `MiscDB` has the db dir of new/old `stateDB`
- remove the trigger in `trie.commit`

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by KimKyungup** [comment]
PTAL. :)

File: storage/database/db_manager.go [hunk 1]
```diff
@@ -402,82 +457,160 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
-		if err := os.RemoveAll(newDBDir); err != nil {
-			logger.Error("Failed to remove the state trie migration database due to an error", "err", err, "dir", newDBDir)
-		} else {
-			logger.Error("Successfully removed the state trie migration database", "dir", newDBDir)
-		}
-		logger.Crit("Killed the process due to the failure of storing the migration status")
+	// Create a new database for migration process.
+	newDB, newDBDir := newStateTrieMigrationDB(dbm.config, blockNum)
+
+	// Store the directory
```
Comment (winnie-byun) [inline storage/database/db_manager.go:561]:
```suggestion
	// Store migration db path in misc db
```

File: storage/database/db_manager.go [hunk 2]
```diff
@@ -402,82 +457,160 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
+func (dbm *databaseManager) getOldStateTrieDBDir() string {
+	// get old DB dir
+	oldDBDir := dbm.getDBDir(StateTrieDB)
+	if oldDBDir == "" {
+		oldDBDir = dbDirs[StateTrieDB]
+	}
+	return oldDBDir
+}
+
+// FinishStateMigration updates stateTrieDB and remove old one.
```
Comment (winnie-byun) [inline storage/database/db_manager.go:580]:
```suggestion
// FinishStateMigration updates stateTrieDB and removes the old one.
// The function should be called only after when state trie migration is finished.
```

File: storage/database/db_manager.go [hunk 3]
```diff
@@ -402,82 +457,160 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
+func (dbm *databaseManager) FinishStateMigration() {
+	oldDB := dbm.dbs[StateTrieDB]
+	newDB := dbm.dbs[StateTrieMigrationDB]
+	oldDBDir := dbm.getOldStateTrieDBDir()
+	newDBDir := dbm.getDBDir(StateTrieMigrationDB)
+
+	dbm.setDBDir(StateTrieDB, newDBDir)
```
Comment (winnie-byun) [inline storage/database/db_manager.go:587]:
```suggestion
	// replace StateTrieDB with new one
	dbm.setDBDir(StateTrieDB, newDBDir)
```

File: storage/database/db_manager.go [hunk 4]
```diff
@@ -402,82 +457,160 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
+	return oldDBDir
+}
+
+// FinishStateMigration updates stateTrieDB and remove old one.
+func (dbm *databaseManager) FinishStateMigration() {
+	oldDB := dbm.dbs[StateTrieDB]
+	newDB := dbm.dbs[StateTrieMigrationDB]
+	oldDBDir := dbm.getOldStateTrieDBDir()
+	newDBDir := dbm.getDBDir(StateTrieMigrationDB)
+
```
Comment (winnie-byun) [inline storage/database/db_manager.go:586]:
```suggestion

	// Replace StateTrieDB with new one
```

File: storage/database/db_manager.go [hunk 5]
```diff
@@ -111,6 +116,14 @@ type DBManager interface {
 	HasStateTrieNode(key []byte) (bool, error)
 	ReadPreimage(hash common.Hash) []byte
 
+	// Read StateTrie from new DB
+	ReadCachedTrieNodeFromNew(hash common.Hash) ([]byte, error)
+	ReadCachedTrieNodePreimageFromNew(secureKey []byte) ([]byte, error)
+	ReadStateTrieNodeFromNew(key []byte) ([]byte, error)
+	HasStateTrieNodeFromNew(key []byte) (bool, error)
+	ReadPreimageFromNew(hash common.Hash) []byte
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:124]: Can you put a more precise adjective instead of `new`? Every new thing finally gets old.

Comment (KimKyungup) [reply storage/database/db_manager.go:124]: I can't find some good adjective. Please recommend good adjective. :)
I will change the adjective word in other PR.

Comment (KimKyungup) [reply storage/database/db_manager.go:124]: As we talked, I will reflect like below. :)
```
ReadCachedTrieNodeFromNew -> ReadCachedTrieNodeFromMigrationDB
```

File: storage/database/db_manager.go [hunk 6]
```diff
@@ -246,15 +261,19 @@ func checkDBEntryConfigRatio() {
...
 	newDBC.LevelDBCacheSize = originalDBC.LevelDBCacheSize * ratio / 100
 	newDBC.OpenFilesLimit = originalDBC.OpenFilesLimit * ratio / 100
 
 	// Update dir to each Database specific directory.
-	newDBC.Dir = filepath.Join(originalDBC.Dir, dbDirs[i])
+	if dbDir == "" {
+		newDBC.Dir = filepath.Join(originalDBC.Dir, dbDirs[i])
+	} else {
+		newDBC.Dir = filepath.Join(originalDBC.Dir, dbDir)
+	}
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:276]: We call call `getDBEntryConfig` with `dbDirs[i]` value. 
In that case, we don't need this if branch.

File: storage/database/db_manager.go [hunk 7]
```diff
@@ -315,20 +337,48 @@ func singleDatabaseDBManager(dbc *DBConfig) (DBManager, error) {
 	return dbm, nil
 }
 
+// getMiscDB returns misc DBManager
+func getMiscDB(dbc *DBConfig) Database {
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:341]: How about changing this function name to `newMiscDB`.
It always creates a new database.

File: storage/database/db_manager.go [hunk 8]
```diff
@@ -315,20 +337,48 @@ func singleDatabaseDBManager(dbc *DBConfig) (DBManager, error) {
...
+	newDBC := getDBEntryConfig(dbc, MiscDB, "")
+	db, err := newDatabase(newDBC, MiscDB)
+
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:344]:
```suggestion
```

File: storage/database/db_manager.go [hunk 9]
```diff
@@ -315,20 +337,48 @@ func singleDatabaseDBManager(dbc *DBConfig) (DBManager, error) {
...
 	dbm := newDatabaseManager(dbc)
 	var db Database
 	var err error
-	for et := 0; et < int(databaseEntryTypeSize); et++ {
-		entryType := DBEntryType(et)
 
-		newDBC := getDBEntryConfig(dbc, entryType)
+	// Create Misc DB first to get the DB directory of stateTrieDB.
+	miscDB := getMiscDB(dbc)
+	miscDB.Meter(dbMetricPrefix + dbDirs[MiscDB] + "/") // Each partition collects metrics independently.
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:360]: This line may be called inside of `getMiscDB`.

File: storage/database/db_manager.go [hunk 10]
```diff
@@ -402,82 +456,162 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
+func (dbm *databaseManager) getDBDir(dbEntry DBEntryType) string {
+	miscDB := dbm.getDatabase(MiscDB)
+	enc, _ := miscDB.Get(databaseDirKey(uint64(dbEntry)))
+	return string(enc)
+}
+
+func (dbm *databaseManager) setDBDir(dbEntry DBEntryType, newDBDir string) {
+	miscDB := dbm.getDatabase(MiscDB)
+	if err := miscDB.Put(databaseDirKey(uint64(dbEntry)), []byte(newDBDir)); err != nil {
+		logger.Crit("Failed to DB dir", "err", err)
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:487]:
```suggestion
		logger.Crit("Failed to put DB dir", "err", err)
```

File: storage/database/db_manager.go [hunk 11]
```diff
@@ -383,9 +429,14 @@ func NewDBManager(dbc *DBConfig) DBManager {
...
-		if inMigration, migrationDir := dbm.getStateTrieMigrationInfo(); inMigration {
-			dbm.newStateTrieDB = createStateTrieMigrationDB(dbm.config, migrationDir)
+		if migrationBlockNum := dbm.getStateTrieMigrationInfo(); migrationBlockNum > 0 {
+			mdb := dbm.getDatabase(StateTrieMigrationDB)
+			if mdb == nil {
+				logger.Error("Failed to load StateTrieMigrationDB database")
+
+			}
 			dbm.inMigration = true
+			dbm.migrationBlockNumber = migrationBlockNum
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:439]: These lines should be the inside of the above if branch.

File: storage/database/db_manager.go [hunk 12]
```diff
@@ -1167,8 +1288,11 @@ func (dbm *databaseManager) WriteMerkleProof(key, value []byte) {
 
 // Cached Trie Node operation.
 func (dbm *databaseManager) ReadCachedTrieNode(hash common.Hash) ([]byte, error) {
+	dbm.lockInMigration.RLock()
+	defer dbm.lockInMigration.RUnlock()
+
 	if dbm.inMigration {
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:1294]: You better use `dbm.InMigration()`.
This comment is for all places using `.inMigration`.

File: storage/database/db_manager.go [hunk 13]
```diff
@@ -246,15 +261,15 @@ func checkDBEntryConfigRatio() {
 
 // getDBEntryConfig returns a new DBConfig with original DBConfig and DBEntryType.
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:262]:
```suggestion
// getDBEntryConfig returns a new DBConfig with original DBConfig, DBEntryType, and dbDir.
```

File: storage/database/db_manager.go [hunk 14]
```diff
@@ -402,82 +453,155 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
+	dbDir := dbDirs[StateTrieMigrationDB] + strconv.FormatUint(blockNum, 10)
+	newDBConfig := getDBEntryConfig(dbc, StateTrieMigrationDB, dbDir)
 	var newDB Database
 	var err error
 	if newDBConfig.NumStateTriePartitions > 1 {
-		newDB, err = newPartitionedDB(newDBConfig, StateTrieDB, newDBConfig.NumStateTriePartitions)
+		newDB, err = newPartitionedDB(newDBConfig, StateTrieMigrationDB, newDBConfig.NumStateTriePartitions)
 	} else {
-		newDB, err = newDatabase(newDBConfig, StateTrieDB)
+		newDB, err = newDatabase(newDBConfig, StateTrieMigrationDB)
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:525]: Currently, we don't support the migration of non-partitioned DB.

File: storage/database/db_manager.go (hunk 12 above)
Comment (KimKyungup) [reply storage/database/db_manager.go:1295]: This code is intended. 
This `RLock` can prevents  to access removed `MigrationDB`.

File: storage/database/db_manager.go [hunk 15]
```diff
@@ -402,82 +453,158 @@ func (dbm *databaseManager) IsPartitioned() bool {
...
+
+// FinishStateMigration updates stateTrieDB and removes the old one.
+// The function should be called only after when state trie migration is finished.
+func (dbm *databaseManager) FinishStateMigration() {
+	oldDB := dbm.dbs[StateTrieDB]
+	newDB := dbm.dbs[StateTrieMigrationDB]
+	oldDBDir := dbm.getOldStateTrieDBDir()
+	newDBDir := dbm.getDBDir(StateTrieMigrationDB)
+
+	// Replace StateTrieDB with new one
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:579]:
```suggestion
```

File: storage/database/db_manager.go [hunk 16]
```diff
@@ -315,20 +333,48 @@ func singleDatabaseDBManager(dbc *DBConfig) (DBManager, error) {
 	return dbm, nil
 }
 
+// newMiscDB creates misc DBManager.
```
Comment (winnie-byun) [inline storage/database/db_manager.go:336]:
```suggestion
// newMiscDB returns misc DBManager. If not exist, the function create DB before returning.
```

File: storage/database/db_manager.go [hunk 17]
```diff
@@ -244,17 +259,17 @@ func checkDBEntryConfigRatio() {
 	}
 }
 
-// getDBEntryConfig returns a new DBConfig with original DBConfig and DBEntryType.
+// getDBEntryConfig returns a new DBConfig with original DBConfig, DBEntryType, and dbDir.
```
Comment (winnie-byun) [inline storage/database/db_manager.go:262]:
```suggestion
// getDBEntryConfig returns a new DBConfig with original DBConfig, DBEntryType and dbDir.
```

**Pull request merged**